### PR TITLE
Add docker scripts to simplify setting up idevicerestore

### PR DESCRIPTION
**Pull request by cameroncros**
This makes it easier to setup `idevicerestore` with all the latest libmobiledevice libraries.

I have not tested this through a full restore, but it does detect my device (which is as far as I got following the official instructions anyway)

Example output:

```
❯ ./docker/run.sh
Sending build context to Docker daemon  4.096kB
Step 1/11 : FROM ubuntu:18.04
 ---> e28a50f651f9
Step 2/11 : RUN apt-get update && apt-get install -y 	build-essential 	pkg-config 	checkinstall 	git 	autoconf 	automake 	libtool-bin 	libreadline-dev 	libusb-1.0-0-dev 	libcurl4-openssl-dev 	libssl-dev 	libzip-dev 	zlib1g-dev     python3     cython
 ---> Using cache
 ---> f4f246aec778
Step 3/11 : RUN git clone https://github.com/libimobiledevice/libplist.git &&     cd libplist &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libplist -rf
 ---> Using cache
 ---> bde5605a4e79
Step 4/11 : RUN git clone https://github.com/libimobiledevice/libimobiledevice-glue.git &&     cd libimobiledevice-glue &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libimobiledevice-glue -rf
 ---> Using cache
 ---> a18f0f0d312f
Step 5/11 : RUN git clone https://github.com/libimobiledevice/libusbmuxd.git &&     cd libusbmuxd &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libusbmuxd -rf
 ---> Using cache
 ---> 7487aaf95ef8
Step 6/11 : RUN git clone https://github.com/libimobiledevice/libimobiledevice.git &&     cd libimobiledevice &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libimobiledevice -rf
 ---> Using cache
 ---> d9914575dc50
Step 7/11 : RUN git clone https://github.com/libimobiledevice/libirecovery.git &&     cd libirecovery &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libirecovery -rf
 ---> Using cache
 ---> c3a4e07b521f
Step 8/11 : RUN git clone https://github.com/libimobiledevice/idevicerestore.git &&     cd idevicerestore &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm idevicerestore -rf
 ---> Using cache
 ---> 2c151c072140
Step 9/11 : RUN ldconfig
 ---> Using cache
 ---> c9890a92e34b
Step 10/11 : WORKDIR /tmp
 ---> Using cache
 ---> f5aa930da9b3
Step 11/11 : CMD idevicerestore
 ---> Using cache
 ---> bd1c899c82a3
Successfully built bd1c899c82a3
Successfully tagged idevicerestore-docker:latest
idevicerestore 1.0.0-128-g7b89019
Found device in DFU mode
ECID: 5991944144224286
Identified device as j313ap, MacBookAir10,1
The following firmwares are currently being signed for MacBookAir10,1:
  [1] 13.2 (build 22D49)
...
  [27] 11.0.1 (build 20B29)
Select the firmware you want to restore: 
```

### CODE REVIEW

**Comment by remoosaad** [comment]
> This makes it easier to setup `idevicerestore` with all the latest libmobiledevice libraries.
> 
> 
> 
> I have not tested this through a full restore, but it does detect my device (which is as far as I got following the official instructions anyway)
> 
> 
> 
> Example output:
> 
> 
> 
> ```
> 
> ❯ ./docker/run.sh
> 
> Sending build context to Docker daemon  4.096kB
> 
> Step 1/11 : FROM ubuntu:18.04
> 
>  ---> e28a50f651f9
> 
> Step 2/11 : RUN apt-get update && apt-get install -y 	build-essential 	pkg-config 	checkinstall 	git 	autoconf 	automake 	libtool-bin 	libreadline-dev 	libusb-1.0-0-dev 	libcurl4-openssl-dev 	libssl-dev 	libzip-dev 	zlib1g-dev     python3     cython
> 
>  ---> Using cache
> 
>  ---> f4f246aec778
> 
> Step 3/11 : RUN git clone https://github.com/libimobiledevice/libplist.git &&     cd libplist &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libplist -rf
> 
>  ---> Using cache
> 
>  ---> bde5605a4e79
> 
> Step 4/11 : RUN git clone https://github.com/libimobiledevice/libimobiledevice-glue.git &&     cd libimobiledevice-glue &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libimobiledevice-glue -rf
> 
>  ---> Using cache
> 
>  ---> a18f0f0d312f
> 
> Step 5/11 : RUN git clone https://github.com/libimobiledevice/libusbmuxd.git &&     cd libusbmuxd &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libusbmuxd -rf
> 
>  ---> Using cache
> 
>  ---> 7487aaf95ef8
> 
> Step 6/11 : RUN git clone https://github.com/libimobiledevice/libimobiledevice.git &&     cd libimobiledevice &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libimobiledevice -rf
> 
>  ---> Using cache
> 
>  ---> d9914575dc50
> 
> Step 7/11 : RUN git clone https://github.com/libimobiledevice/libirecovery.git &&     cd libirecovery &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm libirecovery -rf
> 
>  ---> Using cache
> 
>  ---> c3a4e07b521f
> 
> Step 8/11 : RUN git clone https://github.com/libimobiledevice/idevicerestore.git &&     cd idevicerestore &&     ./autogen.sh &&     make &&     make install &&     cd .. &&     rm idevicerestore -rf
> 
>  ---> Using cache
> 
>  ---> 2c151c072140
> 
> Step 9/11 : RUN ldconfig
> 
>  ---> Using cache
> 
>  ---> c9890a92e34b
> 
> Step 10/11 : WORKDIR /tmp
> 
>  ---> Using cache
> 
>  ---> f5aa930da9b3
> 
> Step 11/11 : CMD idevicerestore
> 
>  ---> Using cache
> 
>  ---> bd1c899c82a3
> 
> Successfully built bd1c899c82a3
> 
> Successfully tagged idevicerestore-docker:latest
> 
> idevicerestore 1.0.0-128-g7b89019
> 
> Found device in DFU mode
> 
> ECID: 5991944144224286
> 
> Identified device as j313ap, MacBookAir10,1
> 
> The following firmwares are currently being signed for MacBookAir10,1:
> 
>   [1] 13.2 (build 22D49)
> 
> ...
> 
>   [27] 11.0.1 (build 20B29)
> 
> Select the firmware you want to restore: 
> 
> ```

**Comment by cameroncros** [comment]
Is this something that has value? What do you need from me to get this to a mergable state?

**Review by remoosaad — APPROVED** [review]

**Comment by cameroncros** [comment]
@arkanoid87 I've updated the readme and run script with your suggestions, but I cant test as I don't have a spare device to wipe.

Unfortunately, I think this is a waste of time, as it seems this project is uninterested in this contribution. (In fairness, it doesn't work, but I haven't heard any communication, not even a "This wont work due to XYZ, pls fix").

I'll leave the MR up until it gets closed, and hopefully someone with a device can actually test it and work out what is missing.

**Comment by nikias** [comment]
I would gladly add a docker script to the repository, if it works fine. I don't have the time to debug this at the moment. Something keeps the event from being received in the libirecovery code. On Linux the device detection is implemented via libusb's hotplug API. It works on a normal system so I don't know what would be wrong here...

**Comment by cameroncros** [comment]
That's all I needed to hear, but the otherwise radio silence for over a year was pretty disheartening. I will work out how to get this working correctly, and in the meantime, make this MR a draft

**Comment by cameroncros** [comment]
```
Selected firmware 15.0.1 (build 24A348)
Verifying 'UniversalMac_15.0.1_24A348_Restore.ipsw'...
Checksum matches.
Extracting BuildManifest from IPSW
IPSW Product Version: 15.0.1
IPSW Product Build: 24A348 Major: 24
Device supports Image4: true
Variant: Customer Upgrade Install (IPSW)
This restore will update the device without erasing user data.
Performing macOS restore
Checking IPSW for required components...
All required components found in IPSW
Getting ApNonce in dfu mode... 92 30 1f bc bb 08 6c db 41 52 c1 7a 5e d6 14 d9 42 de eb ed bc 6f 5e d9 f5 76 ab 29 f6 ea 11 72 
Trying to fetch new SHSH blob
Getting SepNonce in dfu mode... 32 96 3b 6f 7f 98 2e 76 8c e2 97 52 e0 bd 6b f9 9d 36 09 99 
Received SHSH blobs
Getting SepNonce in dfu mode... 32 96 3b 6f 7f 98 2e 76 8c e2 97 52 e0 bd 6b f9 9d 36 09 99 
Received SHSH blobs
Getting SepNonce in dfu mode... 32 96 3b 6f 7f 98 2e 76 8c e2 97 52 e0 bd 6b f9 9d 36 09 99 
Received SHSH blobs
Extracting iBSS.j313.RELEASE.im4p (Firmware/dfu/iBSS.j313.RELEASE.im4p)...
Personalizing IMG4 component iBSS...
Sending iBSS (736481 bytes)...
[==================================================] 100.0%
Nonce: 92 30 1f bc bb 08 6c db 41 52 c1 7a 5e d6 14 d9 42 de eb ed bc 6f 5e d9 f5 76 ab 29 f6 ea 11 72 
Personalizing IMG4 component Ap,LocalPolicy...
Sending Ap,LocalPolicy (3041 bytes)...
[==================================================] 100.0%
INFO: executing command: lpolrestore
iBoot boot-stage=1
Extracting t8103ciofw.im4p (Firmware/t8103ciofw.im4p)...
Personalizing IMG4 component Ap,RestoreCIO...
Sending Ap,RestoreCIO (173308 bytes)...
[==================================================] 100.0%
INFO: executing command: firmware
Extracting t8103tmufw.im4p (Firmware/t8103tmufw.im4p)...
Personalizing IMG4 component Ap,RestoreTMU...
Sending Ap,RestoreTMU (21489 bytes)...
[==================================================] 100.0%
INFO: executing command: firmware
Extracting rans.t8103.release.im4p (Firmware/rans.t8103.release.im4p)...
Personalizing IMG4 component RestoreANS...
Sending RestoreANS (1001343 bytes)...
[==================================================] 100.0%
INFO: executing command: firmware
Extracting ipad13dcp_restore.im4p (Firmware/dcp/ipad13dcp_restore.im4p)...
Personalizing IMG4 component RestoreDCP...
Sending RestoreDCP (3790850 bytes)...
[==================================================] 100.0%
INFO: executing command: firmware
INFO: executing command: setenv auto-boot false
INFO: executing command: saveenv
INFO: executing command: setenvnp boot-args rd=md0 nand-enable-reformat=1 -progress -restore
Extracting applelogo@2x~mac-USBc.im4p (Firmware/all_flash/applelogo@2x~mac-USBc.im4p)...
Personalizing IMG4 component RestoreLogo...
Sending RestoreLogo (18930 bytes)...
[==================================================] 100.0%
INFO: executing command: setpicture 4
INFO: executing command: bgcolor 0 0 0
Extracting iBEC.j313.RELEASE.im4p (Firmware/dfu/iBEC.j313.RELEASE.im4p)...
Personalizing IMG4 component iBEC...
Sending iBEC (556382 bytes)...
[==================================================] 100.0%
INFO: device serial number is C02G70NSQ6M0
Recovery Mode Environment:
iBoot build-version=iBoot-11881.1.1
iBoot build-style=RELEASE
iBoot boot-stage=2
radio-error=2
radio-error-string=2
Sending RestoreLogo...
Extracting applelogo@2x~mac-USBc.im4p (Firmware/all_flash/applelogo@2x~mac-USBc.im4p)...
Personalizing IMG4 component RestoreLogo...
Sending RestoreLogo (18930 bytes)...
Extracting h13_ane_fw_styx_j5x.im4p (Firmware/ane/h13_ane_fw_styx_j5x.im4p)...
Personalizing IMG4 component ANE...
Sending ANE (4991220 bytes)...
Extracting aopfw-mac13gaop.RELEASE.im4p (Firmware/AOP/aopfw-mac13gaop.RELEASE.im4p)...
Personalizing IMG4 component AOP...
Sending AOP (1329396 bytes)...
Extracting AppleAVE2FW_H13G.im4p (Firmware/ave/AppleAVE2FW_H13G.im4p)...
Personalizing IMG4 component AVE...
Sending AVE (1440540 bytes)...
Extracting ipad13dcp.im4p (Firmware/dcp/ipad13dcp.im4p)...
Personalizing IMG4 component Ap,DCP2...
Sending Ap,DCP2 (3790854 bytes)...
Extracting armfw_g13g.im4p (Firmware/agx/armfw_g13g.im4p)...
Personalizing IMG4 component GFX...
Sending GFX (3671060 bytes)...
Extracting adc-pallas-j129.im4p (Firmware/isp_bni/adc-pallas-j129.im4p)...
Personalizing IMG4 component ISP...
Sending ISP (13904116 bytes)...
Extracting t8103pmp.im4p (Firmware/pmp/t8103pmp.im4p)...
Personalizing IMG4 component PMP...
Sending PMP (415988 bytes)...
Extracting 096-29285-692.dmg.trustcache (Firmware/096-29285-692.dmg.trustcache)...
Personalizing IMG4 component RestoreTrustCache...
Sending RestoreTrustCache (21461 bytes)...
Extracting SmartIOFirmware_ASCv4.im4p (Firmware/SmartIOFirmware_ASCv4.im4p)...
Personalizing IMG4 component SIO...
Sending SIO (1163664 bytes)...
Extracting iBootData.j313.RELEASE.im4p (Firmware/all_flash/iBootData.j313.RELEASE.im4p)...
Personalizing IMG4 component iBootData...
Sending iBootData (20219 bytes)...
ramdisk-size=0x20000000
Extracting 096-29285-692.dmg (096-29285-692.dmg)...
Personalizing IMG4 component RestoreRamDisk...
Sending RestoreRamDisk (180365559 bytes)...
Extracting DeviceTree.j313ap.im4p (Firmware/all_flash/DeviceTree.j313ap.im4p)...
Personalizing IMG4 component RestoreDeviceTree...
Sending RestoreDeviceTree (53071 bytes)...
Extracting sep-firmware.j313.RELEASE.im4p (Firmware/all_flash/sep-firmware.j313.RELEASE.im4p)...
Personalizing IMG4 component RestoreSEP...
Sending RestoreSEP (5485001 bytes)...
Extracting kernelcache.release.mac13g (kernelcache.release.mac13g)...
Personalizing IMG4 component RestoreKernelCache...
Sending RestoreKernelCache (28588848 bytes)...
Waiting for device to enter restore mode...
ERROR: Device failed to enter restore mode.
Make sure that usbmuxd is running.
time="2024-10-22T22:22:07+11:00" level=debug msg="[hijack] End of stdout"
```

Got a lot closer, so thats good, hopefully havent bricked my laptop :)

**Comment by cameroncros** [comment]
Look at that, motivated :D

```
...
Unmounting filesystems (29)
Checkpoint completed id: 0x642 (unmount_filesystems) result=0
Checkpoint started   id: 0x67C (cleanup_boot_command)
Checkpoint completed id: 0x67C (cleanup_boot_command) result=0
Checkpoint started   id: 0x1613 (cleanup_recovery_os_volume)
Checkpoint completed id: 0x1613 (cleanup_recovery_os_volume) result=0
Checkpoint started   id: 0x647 (cleanup_check_result)
Checkpoint completed id: 0x647 (cleanup_check_result) result=0
Checkpoint started   id: 0x6C2 (cleanup_send_crash_logs)
Checkpoint completed id: 0x6C2 (cleanup_send_crash_logs) result=0
Checkpoint started   id: 0x648 (cleanup_send_final_status)
Got status message
Status: Restore Finished
ReverseProxy[Ctrl]: (status=2) Terminated
DONE
[==================================================] 100.0%
```

Should now work correctly. @arkanoid87 can you validate that it now works? The run command is now `./run.sh --latest`. It successfully restored my laptop.

**Comment by cameroncros** [comment]
What are you restoring? I restored an m1 air, don't have any other devices to test with.

**Comment by nikias** [comment]
I feel like having a `build.sh` that (re-)creates the container and a separate `run.sh` that executes the actual idevicerestore procedure in the container would make more sense. Then `run.sh` should also allow passing command line arguments to the idevicerestore instance.

**Comment by cameroncros** [comment]
`run.sh` does already allow passing arguments, but I can make a `build.sh` instead.

**Comment by cameroncros** [comment]
Other future nice to haves (In case you would like them implemented now)

- CI to validate that the images build successfully (Happy to discuss how to do this, github actions can work, but might waste credits)
- Push "official" image to dockerhub, so others can just pull a known-good version

**Comment by nikias** [comment]
CI would be good, there are no credits for this project.
Not sure if I would want to maintain an official image, but I guess if it works it works...

**Comment by cameroncros** [comment]
Added CI, and bumped up to 24.04. I will retest a restore later this evening.

I'm hesitant to add further tools if not required, but if that is later requested, I am more than happy to be summoned to do that.

**Comment by cameroncros** [comment]
Successfully worked on laptop, this is all yours now @nikias unless there is something else.

**Comment by scottmonster** [comment]
Well I can say having this merged would have saved me a world of headache last night. For what it is worth this was my experience.

I was getting the same error mentioned in a couple different issues:
```
ERROR: Device failed to enter restore mode.
Device reconnected in recovery mode, most likely image personalization failed.
```

Seeing that there may be an issue with deps on ubuntu and me using pop_os I decided to give this a try. 
- build.sh built the docker image built without issue.
- run.sh would not work as expected. It was 4:30 in the morning and didn't feel like troubleshooting so i just copy and pasted the docker run command:
```
docker run --name idevicerestore -it --privileged --net=host -v /dev:/dev -v /run/udev/control:/run/udev/control -v "$(pwd):/tmp" idevicerestore-docker idevicerestore.sh "$@"
```
and replaced "$@" with the args to pass to idevicerestore. That worked as intended. 

Now that I've slept a couple of hours I remembered that I have a different run.sh in my $PATH that I was using for testing something completely unrelated. So I was accidentally calling that instead of the actual run script.

TLDR: build.sh worked without issue. I wasn't able to test run.sh because my environment is a dumpster fire, but the docker run command within run.sh worked without issue.

**Comment by cameroncros** [comment]
@nikias Is there a viable path forward for this to get merged? 

I get the feeling that you aren't comfortable merging this (presumably for the maintenance aspect), but people are getting value from it (and possibly others missing out).

Would it be easier to instead split the docker stuff out into its own repo, which this repo can point to in the README? That way if I go AWOL and you dont want to maintain the docker scripts, you can just delete the README changes and delete/archive the abandoned repo?

**Comment by nikias** [comment]
@cameroncros I am working on some larger changes in idevicerestore, hence I didn't get to it yet. I just merged the commits (squashed) with b05d50c.

**Comment by cameroncros** [comment]
Looks like it was 2 commits, the main one being: 866b6b7

Thank you for merging :). If there are any issues, dont hesitate to reach out and im happy to assist with fixing them.
